File: spikes/iac-prototype/iac-copilot-api/models/skill_package.py
"""Models for Skill Packages — real, filesystem-backed skill directories that
conform to the agentskills.io Agent Skills specification. Distinct from the
legacy frontend "Skills" concept and from the `skills/` tool-registry package
used by the agent engagement loop.
"""

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel


class SkillPackageManifest(BaseModel):
    """Matches the SKILL.md YAML frontmatter fields defined by
    https://agentskills.io/specification."""
    name: str
    description: str
    license: Optional[str] = None
    compatibility: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None
    allowed_tools: Optional[str] = None


class InstalledSkillPackage(BaseModel):
    """Lightweight record of a skill package installed on disk. Mirrors what's
    actually at `install_path` — not a cache the frontend should trust over
    the filesystem."""
    name: str
    install_path: str
    installed_at: datetime
    source: Literal["upload", "host"] = "upload"
    source_filename: Optional[str] = None
    sha256: Optional[str] = None
    manifest: SkillPackageManifest
    files: List[str] = []


class SkillPackageInstallResponse(BaseModel):
    skill: InstalledSkillPackage
    warnings: List[str] = []


class SkillPackageListResponse(BaseModel):
    skills: List[InstalledSkillPackage]
    total: int


class SkillPackageUninstallResponse(BaseModel):
    success: bool
    name: str

File: spikes/iac-prototype/iac-host/src/types/skillPackage.ts
/**
 * Types for Skill Packages — real, filesystem-backed skill directories
 * conforming to the agentskills.io Agent Skills specification.
 *
 * Distinct from the legacy `types/skills.ts` (`SkillManifest`/`InstalledSkill`
 * — a JSON manifest wrapping a whole SKILL.md as one string, localStorage
 * only) and from the unrelated backend tool-registry in
 * `iac-copilot-api/skills/`. Field names are snake_case to match the
 * backend's Pydantic JSON output directly, same convention as `types/agents.ts`.
 */

/** Matches the SKILL.md YAML frontmatter fields defined by the spec. */
export interface SkillPackageManifest {
  name: string;
  description: string;
  license?: string;
  compatibility?: string;
  metadata?: Record<string, string>;
  allowed_tools?: string;
}

/** Lightweight pointer to a skill package installed on disk at
 * `install_path` — not the file contents. */
export interface InstalledSkillPackage {
  name: string;
  install_path: string;
  installed_at: string;
  source: 'upload' | 'host';
  source_filename?: string;
  sha256?: string;
  manifest: SkillPackageManifest;
  files: string[];
}

export interface SkillPackageInstallResponse {
  skill: InstalledSkillPackage;
  warnings: string[];
}

export interface SkillPackageListResponse {
  skills: InstalledSkillPackage[];
  total: number;
}

export interface SkillPackageContextValue {
  installed: InstalledSkillPackage[];
  isLoading: boolean;
  isInstalling: boolean;
  error: string | null;
  refreshInstalled: () => Promise<void>;
  installFromFile: (file: File, overwrite?: boolean) => Promise<InstalledSkillPackage>;
  uninstall: (name: string) => Promise<void>;
}
